src/execution/compiler/operator/seq_scan_translator.cpp

```diff
@@ -1,231 +1,344 @@
-  // Let parent consume.
-  parent_translator_->Consume(builder);
-  // Close predicate if statement
-  if (has_if_stmt) {
-    builder->FinishBlockStmt();
+
+  // Consecutive conjunctions are part of the same clause.
+  if (predicate->GetExpressionType() == parser::ExpressionType::CONJUNCTION_AND) {
+    GenerateFilterClauseFunctions(decls, predicate->GetChild(0), curr_clause, true);
+    GenerateFilterClauseFunctions(decls, predicate->GetChild(1), curr_clause, true);
+    return;
   }
-  // Close PCI loop
-  builder->FinishBlockStmt();
-  // Close TVI loop
-  builder->FinishBlockStmt();
-}
 
-void SeqScanTranslator::Consume(FunctionBuilder *builder) {
-  // This is called in nested loop joins
-  DoTableScan(builder);
-  // Reset TVI for next iteration.
-  GenTVIReset(builder);
-}
+  // At this point, we create a term.
+  // Signature: (execCtx: *ExecutionContext, vp: *VectorProjection, tids: *TupleIdList, ctx: *uint8) -> nil
+  auto *codegen = GetCodeGen();
+  auto fn_name = codegen->MakeFreshIdentifier(GetPipeline()->CreatePipelineFunctionName("FilterClause"));
+  util::RegionVector<ast::FieldDecl *> params = codegen->MakeFieldList({
+      codegen->MakeField(codegen->MakeIdentifier("execCtx"), codegen->PointerType(ast::BuiltinType::ExecutionContext)),
+      codegen->MakeField(codegen->MakeIdentifier("vp"), codegen->PointerType(ast::BuiltinType::VectorProjection)),
+      codegen->MakeField(codegen->MakeIdentifier("tids"), codegen->PointerType(ast::BuiltinType::TupleIdList)),
+      codegen->MakeField(codegen->MakeIdentifier("context"), codegen->PointerType(ast::BuiltinType::Uint8)),
+  });
+  FunctionBuilder builder(codegen, fn_name, std::move(params), codegen->Nil());
+  {
+    ast::Expr *exec_ctx = builder.GetParameterByPosition(0);
+    ast::Expr *vector_proj = builder.GetParameterByPosition(1);
+    ast::Expr *tid_list = builder.GetParameterByPosition(2);
+    if (parser::ExpressionUtil::IsColumnCompareWithConst(*predicate)) {
+      auto cve = predicate->GetChild(0).CastManagedPointerTo<parser::ColumnValueExpression>();
+      auto translator = GetCompilationContext()->LookupTranslator(*predicate->GetChild(1));
+      auto col_index = GetColOidIndex(cve->GetColumnOid());
+      auto const_val = translator->DeriveValue(nullptr, nullptr);
+      builder.Append(codegen->VPIFilter(exec_ctx,                        // The execution context
+                                        vector_proj,                     // The vector projection
+                                        predicate->GetExpressionType(),  // Comparison type
+                                        col_index,                       // Column index
+                                        const_val,                       // Constant value
+                                        tid_list));                      // TID list
+    } else if (parser::ExpressionUtil::IsColumnCompareWithParam(*predicate)) {
+      // TODO(WAN): temporary hacky implementation, poke Prashanth...
```

Comment:
Yes, I get the gist. But, just derive the value of the param expression using the translator rather than manually doing it. That's what they're there for.

Reply:
whoa, nice. This collapses the two cases into one.

Reply:
Actually, no. I remember now why this was done. It ends up generating code like

```
fun Query3_Pipeline1_FilterClause1(execCtx: *ExecutionContext, vp: *VectorProjection, tids: *TupleIdList, context: *uint8) -> nil {
    @filterGe(execCtx, vp, 1, @getParamInt(queryState.execCtx, 2), tids)
    return
}
```

Reply:
If this isn't urgent, I would remove it altogether and create an issue to address later.

The correct way to do this is to modify the filter manage to always pass in the query state to filter functions. From that, we can derive the execution context and the parameter. This also allows us to remove the existing `ExecutionContext *` argument, as discussed on Slack.